src/sparse/merkle_tree.rs

```diff
@@ -0,0 +1,445 @@
+    }
+
+    #[test]
+    fn test_update_1_update_empty() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        let key = 0_u32.to_be_bytes();
+        let data = b"DATA";
+        tree.update(&key, data).unwrap();
+        tree.update(&key, &[0; 0]).unwrap();
+
+        let root = tree.root();
+        let expected_root = "0000000000000000000000000000000000000000000000000000000000000000";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_2_delete_1() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        for i in 0_u32..2 {
+            let key = i.to_be_bytes();
+            let data = b"DATA";
+            tree.update(&key, data).unwrap();
+        }
+
+        let key = 0_u32.to_be_bytes();
+        tree.delete(&key).unwrap();
+
+        let root = tree.root();
+        let expected_root = "d7cb6616832899ac111a852ca8df2d63a1cdb36cb84651ffde72e264506a456f";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_10_delete_5() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        for i in 0_u32..10 {
+            let key = i.to_be_bytes();
+            let data = b"DATA";
+            tree.update(&key, data).unwrap();
+        }
+
+        for i in 5_u32..10 {
+            let key = i.to_be_bytes();
+            tree.delete(&key).unwrap();
+        }
+
+        let root = tree.root();
+        let expected_root = "108f731f2414e33ae57e584dc26bd276db07874436b2264ca6e520c658185c6b";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_interleaved_update_delete() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        let data = b"DATA";
+
+        for i in 0_u32..10 {
+            let key = i.to_be_bytes();
+            tree.update(&key, data).unwrap();
+        }
+
+        for i in 5_u32..15 {
+            let key = i.to_be_bytes();
+            tree.delete(&key).unwrap();
+        }
+
+        for i in 10_u32..20 {
+            let key = i.to_be_bytes();
+            tree.update(&key, data).unwrap();
+        }
+
+        for i in 15_u32..25 {
+            let key = i.to_be_bytes();
+            tree.delete(&key).unwrap();
+        }
+
+        for i in 20_u32..30 {
+            let key = i.to_be_bytes();
+            tree.update(&key, data).unwrap();
+        }
+
+        for i in 25_u32..35 {
+            let key = i.to_be_bytes();
+            tree.delete(&key).unwrap();
+        }
+
+        let root = tree.root();
+        let expected_root = "7e6643325042cfe0fc76626c043b97062af51c7e9fc56665f12b479034bce326";
```

Comment:
This test also deviates from the specs

Reply:
Updated the spec

Reply:
Updated test spec is now in master! https://github.com/FuelLabs/fuel-merkle/blob/master/docs/test-specs/sparse_merkle_tree_tests.md#test-interleaved-update-delete